File: VB/DashboardDiagnosticToolUI/DiagnosticForm.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Threading
Imports System.Windows.Forms
Imports DashboardDiagnosticTool
Imports DashboardDiagnosticTool.Data
Imports DevExpress.Utils.About
Imports DevExpress.Utils.CommonDialogs.Internal
Imports DevExpress.XtraBars
Imports DevExpress.XtraEditors
Imports DevExpress.XtraPrinting.Native
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes
Imports System.Runtime.InteropServices

Namespace DashboardDiagnosticToolUI

    Public Partial Class DiagnosticForm
        Inherits XtraForm

        Private ReadOnly Property Sessions As IList(Of SessionItem)
            Get
                Return controller.Sessions
            End Get
        End Property

        Private ReadOnly Property InProcess As Boolean
            Get
                Return controller.InProcess
            End Get
        End Property

        Private processCount As Integer = 0

        Private benchmarkState As Dictionary(Of BenchmarkItem, Boolean) = New Dictionary(Of BenchmarkItem, Boolean)()

        Private controller As DiagnosticController = New DiagnosticController(New FileController())

        Private waitForm As WaitForm1

        Public Sub GetCommand(ByVal command As String, ByVal args As Object())
            Select Case command.ToLower()
                Case "open"
                    controller.Open()
                Case "save"
                    controller.Save()
                Case "saveas"
                    controller.SaveAs()
                Case "start"
                    controller.Start()
                Case "stop"
                    controller.[Stop]()
                Case "delete"
                    controller.Delete(CType(args(0), SessionItem))
                Case "about"
                    AboutHelper.Show(ProductKind.XtraReports, ProductInfoHelper.GetProductInfo(ProductKind.XtraReports))
            End Select
        End Sub

        Public Function CanHandleCommand(ByVal command As String) As Boolean
            Select Case command.ToLower()
                Case "open"
                    Return controller.CanHandleCommand(ControllerCommand.Open)
                Case "save"
                    Return controller.CanHandleCommand(ControllerCommand.Save)
                Case "saveas"
                    Return controller.CanHandleCommand(ControllerCommand.SaveAs)
                Case "start"
                    Return controller.CanHandleCommand(ControllerCommand.Start)
                Case "stop"
                    Return controller.CanHandleCommand(ControllerCommand.[Stop])
                Case "delete"
                    Return controller.CanHandleCommand(ControllerCommand.Delete)
                Case "about"
                    Return True
            End Select

            Return False
        End Function

        Public Sub New()
            InitializeComponent()
            treeListSession.DataSource = Sessions
            Me.controller.ThrowException += AddressOf OnThrowException
            Me.controller.SessionsChanged += AddressOf OnSessionsChanged
            Me.controller.StartProcessing += AddressOf OnStartProcessing
            Me.controller.EndProcessing += AddressOf OnEndProcessing
            UpdateBarItems()
        End Sub

        Private Sub OnStartProcessing()
            Interlocked.Increment(processCount)
            If processCount > 0 Then Return
            BeginMethod(Sub()
                waitForm = New WaitForm1() With {.StartPosition = FormStartPosition.Manual}
                waitForm.Location = GetWaitLocation(waitForm.ClientRectangle)
                waitForm.Show(Me)
            End Sub)
        End Sub

        Private Function GetWaitLocation(ByVal waitRect As Rectangle) As Point
            Dim rect = RectHelper.AlignRectangle(waitRect, ClientRectangle, ContentAlignment.MiddleCenter)
            Return PointToScreen(rect.Location)
        End Function

        Private Sub OnEndProcessing()
            Interlocked.Decrement(processCount)
            If processCount > 0 Then Return
            BeginMethod(Sub()
                waitForm?.Close()
                waitForm = Nothing
            End Sub)
        End Sub

        Private Sub OnThrowException(ByVal e As Exception)
            If InvokeRequired Then
                BeginMethod(New Action(Of Exception)(AddressOf ShowError), e)
            Else
                ShowError(e)
            End If
        End Sub

        Private Sub OnSessionsChanged()
            BeginMethod(Sub()
                treeListSession.RefreshDataSource()
                If InProcess Then FocusNode(treeListSession, Sessions.Count - 1)
                SyncState()
                UpdateBarItems()
            End Sub)
        End Sub

        Private Sub BeginMethod(ByVal method As Action)
            BeginInvoke(method)
        End Sub

        Private Sub BeginMethod(Of T)(ByVal method As Action(Of T), ByVal arg As T)
            Me.BeginInvoke(New Action(Of T)(method), arg)
        End Sub

        Private Sub SyncState()
            Dim todelete = benchmarkState.Keys.Where(Function(x) Not controller.Benchmarks.ContainsKey(x.SessionId)).ToArray()
            For Each item In todelete
                benchmarkState.Remove(item)
            Next
        End Sub

        Private Shared Sub FocusNode(ByVal treeList As TreeList, ByVal id As Integer)
            Dim node = treeList.FindNodeByID(id)
            If node IsNot Nothing Then
                node.ParentNode?.Expand()
                treeList.FocusedNode = node
            End If
        End Sub

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                controller.Dispose()
                components?.Dispose()
                waitForm?.Close()
                waitForm = Nothing
            End If

            MyBase.Dispose(disposing)
        End Sub

        Private Sub UpdateBarItems()
            barItemStatus.Caption = If(InProcess, "In process", "Stopped")
            For Each item In New BarItem() {barItemStop, barItemStart, barItemOpen, barItemDelete, barItemSave, barItemSaveAs}
                item.Enabled = CanHandleCommand(TryCast(item.Tag, String))
            Next
        End Sub

        Private Sub barManager_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            Dim command = TryCast(e.Item.Tag, String)
            If String.IsNullOrEmpty(command) Then Return
            Dim item = TryCast(treeListSession.GetFocusedRow(), SessionItem)
            Me.HandleCommand(command, New Object() {item})
        End Sub

        Private Overloads Sub OnKeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
            Select Case e.KeyCode
                Case Keys.Delete
                    Dim item = TryCast(treeListSession.GetFocusedRow(), SessionItem)
                    Me.HandleCommand("delete", New Object() {item})
                    e.Handled = True
                Case Keys.F5
                    HandleCommand(If(e.Shift, "stop", "start"))
                    e.Handled = True
                Case Keys.S
                    If e.Control Then
                        HandleCommand(If(e.Shift, "saveas", "save"))
                        e.Handled = True
                    End If

                Case Keys.O
                    If e.Control Then
                        HandleCommand("open")
                        e.Handled = True
                    End If

                Case Keys.Right
                    Dim focusedNodeExpand As TreeListNode = treeListBenchmark.FocusedNode
                    If focusedNodeExpand.HasChildren Then focusedNodeExpand.Expand()
                Case Keys.Left
                    Dim focusedNodeCollapse As TreeListNode = treeListBenchmark.FocusedNode
                    focusedNodeCollapse.Collapse()
            End Select
        End Sub

        Private Sub HandleCommand(ByVal command As String, ByVal Optional args As Object() = Nothing)
            Select Case command.ToLower()
                Case "exit"
                    Close()
                    Return
            End Select

            If CanHandleCommand(command) Then
                GetCommand(command, args)
                UpdateBarItems()
            End If
        End Sub

        Private Sub ShowError(ByVal e As Exception)
            XtraMessageBox.Show(LookAndFeel, Me, e.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        End Sub

        Private Sub ShowSession(ByVal session As Integer)
            treeListBenchmark.DataSource = controller.GetBenchmarks(session)
            gridEvent.DataSource = controller.GetTraceEvents(session)
        End Sub

        Private Sub ClearSession()
            treeListBenchmark.DataSource = Nothing
            gridEvent.DataSource = Nothing
        End Sub

        Private Sub treeListSession_FocusedNodeChanged(ByVal sender As Object, ByVal e As FocusedNodeChangedEventArgs)
            Dim item = TryCast(treeListSession.GetFocusedRow(), SessionItem)
            If item IsNot Nothing Then
                Me.IterateBenchmarks(Sub(node, bm) benchmarkState(bm) = node.Expanded)
                ShowSession(item.ID)
                Me.IterateBenchmarks(Sub(node, bm) node.Expanded = Me.GetExpanded(bm))
            Else
                ClearSession()
            End If
        End Sub

        Private Sub treeListBenchmark_FocusedNodeChanged(ByVal sender As Object, ByVal e As FocusedNodeChangedEventArgs)
            Dim item = TryCast(treeListBenchmark.GetFocusedRow(), BenchmarkItem)
            If item IsNot Nothing Then
                gridEvent.DataSource = controller.GetTraceEvents(item.SessionId, item)
            End If
        End Sub

        Private Function GetExpanded(ByVal bm As BenchmarkItem) As Boolean
            Dim value As Boolean
            Return If(benchmarkState.TryGetValue(bm, value), value, False)
        End Function

        Private Sub IterateBenchmarks(ByVal action As Action(Of TreeListNode, BenchmarkItem))
            treeListBenchmark.NodesIterator.[Do](Sub(x)
                Dim item = TryCast(treeListBenchmark.GetDataRecordByNode(x), BenchmarkItem)
                If item IsNot Nothing Then action(x, item)
            End Sub)
        End Sub
    End Class

    Friend Class FileController
        Inherits IFileController

        Const fileExt As String = "xml"

        Public Function TryOpenFile(<Out> ByRef openName As String, ByVal Optional fileName As String = "") As Boolean
            Using dlg = CommonDialogProvider.Instance.CreateDefaultOpenFileDialog()
                dlg.Title = "Open"
                dlg.ValidateNames = True
                dlg.Filter = $"Diagnostic files (*.{fileExt})|*.{fileExt}"
                dlg.FilterIndex = 1
                If dlg.ShowDialog() = DevExpress.Utils.CommonDialogs.Internal.DialogResult.OK Then
                    openName = dlg.FileName
                    Return True
                End If

                openName = ""
                Return False
            End Using
        End Function

        Public Function TrySaveFile(<Out> ByRef outFileName As String, ByVal Optional fileName As String = "") As Boolean
            Using dlg = CommonDialogProvider.Instance.CreateDefaultSaveFileDialog()
                dlg.Title = "Saving"
                dlg.ValidateNames = True
                dlg.Filter = $"Diagnostic files (*.{fileExt})|*.{fileExt}"
                dlg.FilterIndex = 1
                dlg.FileName = fileName
                dlg.OverwritePrompt = True
                If dlg.ShowDialog() = DevExpress.Utils.CommonDialogs.Internal.DialogResult.OK Then
                    outFileName = dlg.FileName
                    Return True
                End If

                outFileName = ""
                Return False
            End Using
        End Function
    End Class
End Namespace

File: DashboardDiagnosticToolUI/DiagnosticForm.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Threading;
using System.Windows.Forms;
using DashboardDiagnosticTool;
using DashboardDiagnosticTool.Data;
using DevExpress.Utils.About;
using DevExpress.Utils.CommonDialogs.Internal;
using DevExpress.XtraBars;
using DevExpress.XtraEditors;
using DevExpress.XtraPrinting.Native;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;

namespace DashboardDiagnosticToolUI {
    public partial class DiagnosticForm : XtraForm {
        IList<SessionItem> Sessions => controller.Sessions;
        bool InProcess => controller.InProcess;
        int processCount = 0;

        Dictionary<BenchmarkItem, bool> benchmarkState = new Dictionary<BenchmarkItem, bool>();
        DiagnosticController controller = new DiagnosticController(new FileController());
        WaitForm1 waitForm;

        public void GetCommand(string command, object[] args) {
            switch(command.ToLower()) {
                case "open":
                    controller.Open();
                    break;
                case "save":
                    controller.Save();
                    break;
                case "saveas":
                    controller.SaveAs();
                    break;
                case "start":
                    controller.Start();
                    break;
                case "stop":
                    controller.Stop();
                    break;
                case "delete":
                    controller.Delete((SessionItem)args[0]);
                    break;
                case "about":
                    AboutHelper.Show(ProductKind.XtraReports, ProductInfoHelper.GetProductInfo(ProductKind.XtraReports));
                    break;
            }
        }

        public bool CanHandleCommand(string command) {
            switch(command.ToLower()) {
                case "open":
                    return controller.CanHandleCommand(ControllerCommand.Open);
                case "save":
                    return controller.CanHandleCommand(ControllerCommand.Save);
                case "saveas":
                    return controller.CanHandleCommand(ControllerCommand.SaveAs);
                case "start":
                    return controller.CanHandleCommand(ControllerCommand.Start);
                case "stop":
                    return controller.CanHandleCommand(ControllerCommand.Stop);
                case "delete":
                    return controller.CanHandleCommand(ControllerCommand.Delete);
                case "about":
                    return true;
            }
            return false;
        }

        public DiagnosticForm() {
            InitializeComponent();
            treeListSession.DataSource = Sessions;
            controller.ThrowException += OnThrowException;
            controller.SessionsChanged += OnSessionsChanged;
            controller.StartProcessing += OnStartProcessing;
            controller.EndProcessing += OnEndProcessing;
            UpdateBarItems();
        }
        void OnStartProcessing() {
            Interlocked.Increment(ref processCount);
            if(processCount > 0) return;
            BeginMethod(() => {
                waitForm = new WaitForm1() { StartPosition = FormStartPosition.Manual };
                waitForm.Location = GetWaitLocation(waitForm.ClientRectangle);
                waitForm.Show(this);
            });
        }
        Point GetWaitLocation(Rectangle waitRect) {
            var rect = RectHelper.AlignRectangle(waitRect, ClientRectangle, DevExpress.XtraPrinting.ImageAlignment.MiddleCenter);
            return PointToScreen(rect.Location);
        }
        void OnEndProcessing() {
            Interlocked.Decrement(ref processCount);
            if(processCount > 0) return;
            BeginMethod(() => {
                waitForm?.Close();
                waitForm = null;
            });
        }
        void OnThrowException(Exception e) {
            if(InvokeRequired)
                BeginMethod(ShowError, e);
            else
                ShowError(e);
        }
        void OnSessionsChanged() {
            BeginMethod(() => {
                treeListSession.RefreshDataSource();
                if(InProcess)
                    FocusNode(treeListSession, Sessions.Count - 1);
                SyncState();
                UpdateBarItems();
            });
        }
        void BeginMethod(Action method) {
            BeginInvoke(method);
        }
        void BeginMethod<T>(Action<T> method, T arg) {
            BeginInvoke(new Action<T>(method), arg);
        }
        void SyncState() {
            var todelete = benchmarkState.Keys.Where(x => !controller.Benchmarks.ContainsKey(x.SessionId)).ToArray();
            foreach(var item in todelete)
                benchmarkState.Remove(item);
        }
        static void FocusNode(TreeList treeList, int id) {
            var node = treeList.FindNodeByID(id);
            if(node != null) {
                node.ParentNode?.Expand();
                treeList.FocusedNode = node;
            }
        }
        protected override void Dispose(bool disposing) {
            if(disposing) {
                controller.Dispose();
                components?.Dispose();
                waitForm?.Close();
                waitForm = null;
            }
            base.Dispose(disposing);
        }
        void UpdateBarItems() {
            barItemStatus.Caption = InProcess ? "In process" : "Stopped";
            foreach(var item in new BarItem[] { barItemStop, barItemStart, barItemOpen, barItemDelete, barItemSave, barItemSaveAs })
                item.Enabled = CanHandleCommand(item.Tag as string);
        }
        void barManager_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            var command = (e.Item.Tag as string);
            if(string.IsNullOrEmpty(command)) return;
            var item = treeListSession.GetFocusedRow() as SessionItem;
            HandleCommand(command, new object[] { item });
        }
        void OnKeyDown(object sender, KeyEventArgs e) {
            switch(e.KeyCode) {
                case Keys.Delete:
                    var item = treeListSession.GetFocusedRow() as SessionItem;
                    HandleCommand("delete", new object[] { item });
                    e.Handled = true;
                    break;
                case Keys.F5:
                    HandleCommand(e.Shift ? "stop" : "start");
                    e.Handled = true;
                    break;
                case Keys.S:
                    if(e.Control) {
                        HandleCommand(e.Shift? "saveas": "save");
                        e.Handled = true;
                    }
                    break;
                case Keys.O:
                    if(e.Control) {
                        HandleCommand("open");
                        e.Handled = true;
                    }
                    break;
                case Keys.Right:
                    TreeListNode focusedNodeExpand = treeListBenchmark.FocusedNode;
                    if (focusedNodeExpand.HasChildren)
                        focusedNodeExpand.Expand();
                    break;
                case Keys.Left:
                    TreeListNode focusedNodeCollapse = treeListBenchmark.FocusedNode;
                    focusedNodeCollapse.Collapse();
                    break;
            }
        }
        void HandleCommand(string command, object[] args = null) {
            switch(command.ToLower()) {
                case "exit":
                    Close();
                    return;
            }
            if(CanHandleCommand(command)) {
                GetCommand(command, args);
                UpdateBarItems();
            }
        }
        void ShowError(Exception e) {
            XtraMessageBox.Show(LookAndFeel, this, e.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
        }
        void ShowSession(int session) {
            treeListBenchmark.DataSource = controller.GetBenchmarks(session);
            gridEvent.DataSource = controller.GetTraceEvents(session);
        }
        void ClearSession() {
            treeListBenchmark.DataSource = null;
            gridEvent.DataSource = null;
        }
        void treeListSession_FocusedNodeChanged(object sender, FocusedNodeChangedEventArgs e) {
            var item = treeListSession.GetFocusedRow() as SessionItem;
            if(item != null) {
                IterateBenchmarks((node, bm) => benchmarkState[bm] = node.Expanded);
                ShowSession(item.ID);
                IterateBenchmarks((node, bm) => node.Expanded = GetExpanded(bm));
            } else {
                ClearSession();
            }
        }

        void treeListBenchmark_FocusedNodeChanged(object sender, FocusedNodeChangedEventArgs e) {
            var item = treeListBenchmark.GetFocusedRow() as BenchmarkItem;
            if(item != null) {
                gridEvent.DataSource = controller.GetTraceEvents(item.SessionId, item);
            }
        }
        bool GetExpanded(BenchmarkItem bm) {
            bool value;
            return benchmarkState.TryGetValue(bm, out value) ? value : false;
        }
        void IterateBenchmarks(Action<TreeListNode, BenchmarkItem> action) {
            treeListBenchmark.NodesIterator.Do(x => {
                var item = treeListBenchmark.GetDataRecordByNode(x) as BenchmarkItem;
                if(item != null) action(x, item);
            });
        }
    }

    class FileController : IFileController {
        const string fileExt = "xml";
        public bool TryOpenFile(out string openName, string fileName = "") {
            using(var dlg = CommonDialogProvider.Instance.CreateDefaultOpenFileDialog()) {
                dlg.Title = "Open";
                dlg.ValidateNames = true;
                dlg.Filter = $"Diagnostic files (*.{fileExt})|*.{fileExt}";
                dlg.FilterIndex = 1;

                if(dlg.ShowDialog() == DevExpress.Utils.CommonDialogs.Internal.DialogResult.OK) {
                    openName = dlg.FileName;
                    return true;
                }
                openName = "";
                return false;
            }
        }
        public bool TrySaveFile(out string outFileName, string fileName = "") {
            using(var dlg = CommonDialogProvider.Instance.CreateDefaultSaveFileDialog()) {
                dlg.Title = "Saving";
                dlg.ValidateNames = true;
                dlg.Filter = $"Diagnostic files (*.{fileExt})|*.{fileExt}";
                dlg.FilterIndex = 1;
                dlg.FileName = fileName;
                dlg.OverwritePrompt = true;

                if(dlg.ShowDialog() == DevExpress.Utils.CommonDialogs.Internal.DialogResult.OK) {
                    outFileName = dlg.FileName;
                    return true;
                }
                outFileName = "";
                return false;
            }
        }
    }
}
